### PR TITLE
Template builder

### PR DESCRIPTION
**Pull request by andreievg**
Requires back end branch in this PR: https://github.com/openmsupply/application-manager-server/pull/455

Finaly..

Oops, didn't expect it to be this big, sowy.

Fixes: #239 
Fixes: https://github.com/openmsupply/application-manager-web-app/issues/531
Fixes: https://github.com/openmsupply/application-manager-web-app/issues/530
Fixes: https://github.com/openmsupply/application-manager-web-app/issues/529
Fixes: https://github.com/openmsupply/application-manager-web-app/issues/527
Fixes: https://github.com/openmsupply/application-manager-web-app/issues/523

This was meant to be hacky template builder implementation to demonstrate functionality to Valerio and help Adam with configurations, but ended up making it pretty decent to make it maintainable and the primary way to configure procedures. Hacky template builder was considered ok to merge since it won't touch any existing functionality.

**How to review ?**

Would be good to check the things that are not template builder related, but otherwise play around with the builder (must be logged in as admin user, you'll see a menu at the top show admin configurations)

**Changes in existing functionality, all affected files are listed below, in the order they appear in the PR changes**

Basically all of the commits apart from `template builder changes`, they are not sequential changes btw, so that one commit doesn't change previous commit:

`package.json`

+ clone-deep (for evaluator gui)
+ react-json-view (to display object properties, i.e. structure, when configuring and testing evaluations)

`semantic.less`

Uncomment popup, it's used to display popup on non editable things, to indicate to user why things can't be changed

`table.overrides`

For template list in admin/templates to have some spacing

`sites.overrides`

* template builder related styles
* extended `flex-row/column` with extra classes based on this format `flex-{row/column}-{justify_content}-{align_items}`

`LookupTables...ListMainMenu.tsx/ListTable.tsx/LookUpMainMenu/`

Since lookup tables entry point is relocated to admin, made sure lookup tables still work

`Admin.tsx/AdminOther.tsx`

Admin routes related (only accessible for admin user)

`PageElements.tsx`

* has 'renderConfigElement` parameter to render configuration control next to element in TemplateBuilder
* map over all elements (including invisible ones), ApplicationViewWrapper would not show invisible ones anyways, and TemplateBuilder needs to show all 

`HistoryPanel.tsx`

* For some reason import wasn't working

`Index.ts/SiteLayout.tsx`

* Remove unused routes
* Introduce admin route and move lookup tables to admin routes

`ApplicationCreate.tsx`

* Remove need to application section (not needed, and added difficulty when working with application response in config application)
* Current user info to be accessed from context upstream
* Rename and export getDefaultValues

`UserSelection.tsx/Login.tsx/UserArea.tsx/UserRegistration.tsx/UserState.tsx/fetchUserInfo.tsx`

Changes here are done because i was adding isAdmin

* Strongly type onLogin
* Add isAdmin and make sure orgList is always passed to onLogin

`getSectionDetails.ts`

* Changed when removing application sections

`bildSectionsStructure.tsx` 

* Renamed for some reason to sectionDetails (I guess to be explicit)

`useCreateApplication.tsx`

* Get user info from context
* Don't add application sections
* Await applicationMutation and wait for results (for template builder config application creation)

`useGetApplicationStructure.tsx`

* Force run creation of full application structure (when structure changed but data doesn't)
* listen to structure changes

`useListTemplate.tsx`

* Noticed that template list wasn't updated in dashboard when changing to user/org with no allowable templates, fixed

`usLoadApplication.tsx`

* Build structure from template sections not from application sections (they are removed)

**Structure of template builder**

15 minute video (as an overview of template builder related code)
https://www.dropbox.com/s/e825qfeeud08rmc/template%20builder%20code%20overview.mov?dl=0

### CODE REVIEW

**Comment by CarlosNZ** [comment]
I'm getting compile errors when trying to run, even after a fresh `yarn install`:

```
ERROR in [at-loader] ./src/containers/TemplateBuilder/evaluatorGui/renderEvaluation.tsx:4:31 
    TS2305: Module '"../../../semanticUiReactReExport"' has no exported member 'Portal'.

ERROR in [at-loader] ./src/containers/TemplateBuilder/shared/Evaluation.tsx:2:26 
    TS7016: Could not find a declaration file for module 'lodash'. '/Users/carl/Github/openmsupply-application-manager/application-manager-web-app/node_modules/lodash/lodash.js' implicitly has an 'any' type.
  Try `npm install @types/lodash` if it exists or add a new declaration (.d.ts) file containing `declare module 'lodash';`
```

**Comment by andreievg** [comment]
> I'm getting compile errors when trying to run, even after a fresh `yarn install`:
> 
> ```
> ERROR in [at-loader] ./src/containers/TemplateBuilder/evaluatorGui/renderEvaluation.tsx:4:31 
>     TS2305: Module '"../../../semanticUiReactReExport"' has no exported member 'Portal'.
> 
> ERROR in [at-loader] ./src/containers/TemplateBuilder/shared/Evaluation.tsx:2:26 
>     TS7016: Could not find a declaration file for module 'lodash'. '/Users/carl/Github/openmsupply-application-manager/application-manager-web-app/node_modules/lodash/lodash.js' implicitly has an 'any' type.
>   Try `npm install @types/lodash` if it exists or add a new declaration (.d.ts) file containing `declare module 'lodash';`
> ```

oops

https://github.com/openmsupply/application-manager-web-app/pull/841/commits/916837517ed17ceb643ca6df541d858a7e038059

File: src/utils/helpers/loginCheck.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 const isLoggedIn = () => {
   const JWT = localStorage.getItem('persistJWT')
-  if (JWT == 'undefined') return false
+  if (JWT == undefined) return false
```
Comment (CarlosNZ) [inline src/utils/helpers/loginCheck.ts:3]: Pretty sure this needs to be how it was. That was a bug that took ages to track down -- turns out that the local storage can't store "undefined" as a value, so it stores the string "undefined".

**Comment by CarlosNZ** [comment]
Hey, still can't get this to work. The login page is stuck in and endless loop where it keeps re-loading over and over. 🤷‍♂️

File: src/contexts/UserState.tsx [hunk 2]
```diff
@@ -91,24 +97,26 @@ export function UserProvider({ children }: UserProviderProps) {
...
+        newIsAdmin: !!isAdmin,
+      })
       dispatch({ type: 'setLoading', isLoading: false })
     }
   }
 
   // Initial check for persisted user in local storage
   const JWT = localStorage.getItem('persistJWT')
-  if (JWT == 'undefined') logout()
+  if (JWT == undefined) logout()
```
Comment (CarlosNZ) [inline src/contexts/UserState.tsx:119]: Yes indeed -- this needs to be changed back to fix the perpetual loop I mentioned in other comment!

**Comment by CarlosNZ** [comment]
UI-wise, I think this needs to follow the same layout style as the list and application, etc. If a container has a maximum width, then that container needs to be centered within the overall "content" area, like so:

Screen Shot 2021-07-16 at 9 40 18 AM

File: src/contexts/UserState.tsx (hunk 2 above)
Comment (andreievg) [reply src/contexts/UserState.tsx]: d140f22a8e2bbdbd6c64ac045290d1b1942610e8

File: src/utils/helpers/loginCheck.ts (hunk 1 above)
Comment (andreievg) [reply src/utils/helpers/loginCheck.ts:3]: d140f22a8e2bbdbd6c64ac045290d1b1942610e8

**Comment by andreievg** [comment]
> UI-wise, I think this needs to follow the same layout style as the list and application, etc. If a container has a maximum width, then that container needs to be centered within the overall "content" area, like so:

I think it's ok to make a note of all of the crappy UI bits, but UI for this is pretty bad, and it would be quite a bit of work to make all alignment consistent etc... I am happy to make changes for any bugs or anything that can cause unnecessary frustration for the user (like deleting sections without warning). Everything else we can kick down the road ?

**Comment by andreievg** [comment]
@CarlosNZ I've added confirmation for section and page deletion as per our convo, but not for templatePermission and templateAction deletion (which I think is pretty obvious ?) dafc593bc268bd37b40c8ecc9a2ac83dfdc67971

File: src/utils/helpers/loginCheck.ts (hunk 1 above)
Comment (CarlosNZ) [reply src/utils/helpers/loginCheck.ts:3]: You changed the one in `UserState` -- also needs changing in `loginCheck.ts`

**Comment by CarlosNZ** [comment]
I'm getting stuck in the perpetual loop on /login page again. You still need to change the "undefined" thing in [loginCheck](https://github.com/openmsupply/application-manager-web-app/pull/841#discussion_r677841935), as per this comment, but even that doesn't fix it now (it used to).

Also -- need to resolve conflicts on this. I suspect there's a bit of a clash with the changes I made for the Org Selector, so careful putting it back together with your changes (i.e. check the org selector still works)

**Comment by andreievg** [comment]
I've updated from master, but it required some changes

* Needed to re-use serial await after application is created
* Remove icon filters
* Make sure application create work (with newly added 'applications' field in load template)


@CarlosNZ 

> I'm getting stuck in the perpetual loop on /login page again. You still need to change the "undefined" thing in loginCheck, as per this comment, but even that doesn't fix it now (it used to).

> Also -- need to resolve conflicts on this. I suspect there's a bit of a clash with the changes I made for the Org Selector, so careful putting it back together with your changes (i.e. check the org selector still works)

https://github.com/openmsupply/application-manager-web-app/pull/841/commits/8de78f6f569916c0323b4c0a96ca610ebe953ed8

Looks like in `UserState.tsx` there was a check for JWT == undefined (must have been an artifact from original template builder comment), after that it worked ok

If you are interested, have a look at `useGetAppklicationSerial.ts` and how it's used (i think the sequential 'look' is easier to follow, when useGetApplicationSerial is used, i think it's worth the extra complexity i useGetApplicationSerial though)

**Comment by CarlosNZ** [comment]
Sweet. Feel free to merge anytime you want, we can make small tweaks once it's in master.

Couple of little things I noticed though if you want to address them now:

Screen Shot 2021-08-02 at 12 51 16 PM
Shouldn't say Permission name here

Screen Shot 2021-08-02 at 12 54 37 PM
The link on "Colour" goes to Semantic Icon page, was this intentional?
(And also can we spell "Colour" the NZ way 😉 )

**Review by CarlosNZ — APPROVED** [review]

**Comment by andreievg** [comment]
> Sweet. Feel free to merge anytime you want, we can make small tweaks once it's in master.
> 
> Couple of little things I noticed though if you want to address them now:
> 
> Screen Shot 2021-08-02 at 12 51 16 PM
> Shouldn't say Permission name here
> 
> Screen Shot 2021-08-02 at 12 54 37 PM
> The link on "Colour" goes to Semantic Icon page, was this intentional?
> (And also can we spell "Colour" the NZ way 😉 )

https://github.com/openmsupply/application-manager-web-app/pull/841/commits/c365425bdeada37cca4d289909f1a30ea5c19de2

**Pull request merged**